File: 1748-best-team-with-no-conflicts/best-team-with-no-conflicts.java
class Solution {
    public int bestTeamScore(int[] scores, int[] ages) {
        int n=scores.length;
        int[][] arr=new int[n][2];
        for(int i=0;i<n;i++){
            arr[i][0]=ages[i];
            arr[i][1]=scores[i];
        }
        Arrays.sort(arr,(a,b)->{
            if(a[0]==b[0]){
                return a[1]-b[1];
            }
            return a[0]-b[0];
        });

        int[][] dp=new int[scores.length][scores.length];
        for(int[] a:dp){
            Arrays.fill(a,-1);
        }
        return max(arr,0,dp,-1);

    }
    public int max(int[][] arr,int idx,int[][] dp,int j){
            if(idx>=arr.length){
                return 0;
            }
            if(dp[j+1][idx]!=-1){
                return dp[j+1][idx];
            }
            int pick=0;
            if(j==-1 || arr[j][1]<=arr[idx][1]){//maximizing our score by checking that agr prev curr wala chhota rha apne prev se to skip kr denge beacuse of conflict
                pick=arr[idx][1]+max(arr,idx+1,dp,idx);
            }
            int npick=max(arr,idx+1,dp,j);
            return dp[j+1][idx]=Math.max(pick,npick);
      }
}

File: 1748-best-team-with-no-conflicts/best-team-with-no-conflicts.cpp
class Solution {
public:
    int bestTeamScore(vector<int>& scores, vector<int>& ages) {
        vector<pair<int,int>> arr;
        for(int i=0;i<scores.size();i++){
            arr.push_back({ages[i],scores[i]});
        }
        sort(arr.begin(),arr.end(),[](const pair<int,int> &a,const pair<int,int> &b){
            if(a.first==b.first){
                return b.second>a.second;
            }
            return a.first<b.first;

        });
        vector<vector<long long>> dp(scores.size(),vector<long long>(scores.size()+1,-1));
        return solve(arr,0,-1,dp);
    }
    long long solve(vector<pair<int,int>> &arr,int idx,int prev,vector<vector<long long>> &dp){
        if(idx==arr.size()){
            return 0;
        }
        if(dp[idx][prev+1]!=-1){
            return dp[idx][prev+1];
        }
        long long take=0;
        if(prev==-1 || arr[prev].second<=arr[idx].second){
            take=arr[idx].second+solve(arr,idx+1,idx,dp);
        }
        long long skip=solve(arr,idx+1,prev,dp);
        return dp[idx][prev+1]=max(take,skip);
    }
};